README.md
# Express.js Sessions Store Abstraction Library

 * Last tests against master on CircleCI: [![CircleCI](https://circleci.com/gh/faust64/somesession.svg?style=svg)](https://circleci.com/gh/faust64/somesession)

 * Install with: `npm install somesession`

 * Beware `somesession` would only pull all its depdencies if installed with its `devDependencies`.

Considering production deployments, you would have to pick from these `devDependencies`, those you would actually rely on.
Include them from whatever parent project or other piece of code you'ld be loading `somesession` from.

 * Write backend-agnostic code:

```
// cassandra
const storeOptions = { driver: 'cassandra', database: 'sessions', username: 'casuser', password: 'caspass', host: '3.4.5.6', table: 'sessions' };
// mysql
const storeOptions = ({ driver: 'mysql', database: 'sessions', username: 'myuser', password: 'mypass', host: '2.3.4.5', table: 'sessions' };
// postgres
const storeOptions = ({ driver: 'postgres', database: 'sessions', username: 'pguser', password: 'pgpass', host: '1.2.3.4', table: 'sessions' };
// redis
const storeOptions = ({ driver: 'redis', database: 0, password: 'redisauthpass', host: '4.5.6.7' };
// defaults to sqlite
const storeOptions = ({ database: 'sessions.sqlite', path: './', table: 'sessions' };

const bodyParser = require('body-parser');
const express = require('express');
const http = require('http');
const session = require('express-session');
const store = require('somesession');

const app = express();
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));

const storage = store(storeOptions, session);
app.use(session({
	cookie: { secure: process.env.NODE_ENV === 'production' },
	store: storage, secret: 'secr3t',
	resave: false, saveUninitialized: false,
	cookie: { maxAge: 86400 }
    });

const httpServer = http.createServer(app);

app.get('/', (req, res) => {
	if (req.session !== undefined && req.session.userid !== undefined && req.session.userid !== false) {
	    res.send(req.session.userid);
	} else { res.send('unauthenticated'); }
    });

app.post('/login', (req, res) => {
	let userId = req.body.userId || false;
	if (userId !== false) { req.session.userid = userId; }
	res.redirect('/');
    });

app.post('/logout', (req, res) => {
	if (req.session !== undefined && req.session.userid !== undefined && req.session.userid !== false) {
	    req.session.userid = false;
	    req.session.destroy();
	    delete req.session;
	}
	res.redirect('/');
    });

port = port || 8080;
httpServer.listen(port, '127.0.0.1', (err, res) => {
	if (err) { reject(err); }
	console.log(`mock server listening on 127.0.0.1:${port}`);
	resolve(httpServer);
    });
```


CHANGELOG.md
## v0.0.1 - 201801101221

 - initial release


tests/03-mysql.js
const Promise = require('bluebird');
const store = require('../index.js');
const axios = require('axios');

describe('mysql sessions store', () => {
	if (process.env.TEST_MYSQL) {
	    it('tests mysql driver', (done) => {
		    let express = require('./commons/express-app.js');
		    let bind = require('./commons/express-bind.js');
		    let opts = {
			    database: process.env.MYSQL_DATABASE || 'circleci',
			    driver: 'mysql',
			    host: process.env.MYSQL_HOST || 'localhost',
			    table: process.env.MYSQL_TABLE || 'sessions',
			    port: process.env.MYSQL_PORT || 3306
			};
		    if (process.env.MYSQL_PASSWORD !== undefined) { opts.password = process.env.MYSQL_PASSWORD; }
		    if (process.env.MYSQL_USER !== undefined) { opts.username = process.env.MYSQL_USER; }
		    let storage = store(opts, express.session);
		    express.app.use(express.session({
			    cookie: { secure: process.env.NODE_ENV === 'production' },
			    store: storage, secret: 'secr3t',
			    resave: false, saveUninitialized: false,
			    cookie: { maxAge: 86400 }
			}));
		    let cookie = false;
		    console.log('ready');
		    bind(express.app, 8080)
			.then((ret) => {
				httpHandler = ret;
				return axios.get('http://localhost:8080/');
			    })
			.then((data) => {
				if (data.data !== 'unauthenticated') {
				    console.log(data.data || 'data empty');
				    throw new Error(`unexpected payload querying service with no session yet`);
				}
				return axios.post('http://localhost:8080/login', { userId: 'totopouet' }, { maxRedirects: 0 })
				    .then((e) => done(e || 'unexpected error'))
				    .catch((resp) => {
					    if (resp.response === undefined || resp.response.headers === undefined || resp.response.headers['set-cookie'] === undefined) {
						console.log(resp.response || 'response empty');
						throw new Error(`login did not return with session cookie`);
					    }
					    cookie = { headers: { 'Cookie': resp.response.headers['set-cookie'], withCredentials: true } };
					    console.log('get cookie ' + cookie.headers['Cookie']);
					    return axios.get('http://localhost:8080/', cookie);
					})
				    .then((resp) => {
					    if (resp.data !== 'totopouet') {
						console.log(resp.data);
						throw new Error(`unexpected payload querying service having logged in`);
					    }
					    return axios.post('http://localhost:8080/logout', cookie, { maxRedirects: 0 })
						.then((e) => done(e || 'unexpected error'))
						.catch((conf) => {
							httpHandler.close();
							if (conf.response === undefined || conf.response.headers === undefined) {
							    done(`headers missing serving logout`);
							} else if (conf.response.headers['set-cookie'] !== undefined) {
							    done(`session still shows after logout`);
							} else { done(); }
						    });
					})
				    .catch((e) => done(e || `undefined error confirming login succeeded`));
			    })
			.catch((e) => done(e || `undefined error fetching app index`));
		});
	}
    });


tests/04-redis.js
const Promise = require('bluebird');
const store = require('../index.js');
const axios = require('axios');

describe('redis sessions store', () => {
	if (process.env.TEST_REDIS) {
	    it('tests redis driver', (done) => {
		    let express = require('./commons/express-app.js');
		    let bind = require('./commons/express-bind.js');
		    let opts = {
			    database: process.env.REDIS_DB || 0,
			    driver: 'redis',
			    host: '127.0.0.1',
			    port: process.env.REDIS_PORT || 6379
			};
		    if (process.env.REDIS_PASS !== undefined) { opts.password = process.env.REDIS_PASS; }
		    let storage = store(opts, express.session);
		    express.app.use(express.session({
			    cookie: { secure: process.env.NODE_ENV === 'production' },
			    store: storage, secret: 'secr3t',
			    resave: false, saveUninitialized: false,
			    cookie: { maxAge: 86400 }
			}));
		    let cookie = false;
		    console.log('ready');
		    bind(express.app, 8080)
			.then((ret) => {
				httpHandler = ret;
				return axios.get('http://localhost:8080/');
			    })
			.then((data) => {
				if (data.data !== 'unauthenticated') {
				    console.log(data.data || 'data empty');
				    throw new Error(`unexpected payload querying service with no session yet`);
				}
				return axios.post('http://localhost:8080/login', { userId: 'totopouet' }, { maxRedirects: 0 })
				    .then((e) => done(e || 'unexpected error'))
				    .catch((resp) => {
					    if (resp.response === undefined || resp.response.headers === undefined || resp.response.headers['set-cookie'] === undefined) {
						console.log(resp.response || 'response empty');
						throw new Error(`login did not return with session cookie`);
					    }
					    cookie = { headers: { 'Cookie': resp.response.headers['set-cookie'], withCredentials: true } };
					    console.log('get cookie ' + cookie.headers['Cookie']);
					    return axios.get('http://localhost:8080/', cookie);
					})
				    .then((resp) => {
					    if (resp.data !== 'totopouet') {
						console.log(resp.data);
						throw new Error(`unexpected payload querying service having logged in`);
					    }
					    return axios.post('http://localhost:8080/logout', cookie, { maxRedirects: 0 })
						.then((e) => done(e || 'unexpected error'))
						.catch((conf) => {
							httpHandler.close();
							if (conf.response === undefined || conf.response.headers === undefined) {
							    done(`headers missing serving logout`);
							} else if (conf.response.headers['set-cookie'] !== undefined) {
							    done(`session still shows after logout`);
							} else { done(); }
						    });
					})
				    .catch((e) => done(e || `undefined error confirming login succeeded`));
			    })
			.catch((e) => done(e || `undefined error fetching app index`));
		});
	}
    });


tests/01-sqlite.js
const Promise = require('bluebird');
const store = require('../index.js');
const axios = require('axios');

describe('sqlite sessions store', () => {
	if (process.env.TEST_SQLITE) {
	    it('tests sqlite driver', (done) => {
		    let express = require('./commons/express-app.js');
		    let bind = require('./commons/express-bind.js');
		    let opts = {
			    database: 'test.sqlite',
			    driver: 'sqlite',
			    path: './',
			    table: 'sessions'
			};
		    let storage = store(opts, express.session);
		    express.app.use(express.session({
			    cookie: { secure: process.env.NODE_ENV === 'production' },
			    store: storage, secret: 'secr3t',
			    resave: false, saveUninitialized: false,
			    cookie: { maxAge: 86400 }
			}));
		    let cookie = false, httpHandler = false;
		    bind(express.app, 8080)
			.then((ret) => {
				httpHandler = ret;
				return axios.get('http://localhost:8080/');
			    })
			.then((data) => {
				if (data.data !== 'unauthenticated') {
				    console.log(data.data || 'data empty');
				    throw new Error(`unexpected payload querying service with no session yet`);
				}
				return axios.post('http://localhost:8080/login', { userId: 'totopouet' }, { maxRedirects: 0 })
				    .then((e) => done(e || 'unexpected error'))
				    .catch((resp) => {
					    if (resp.response === undefined || resp.response.headers === undefined || resp.response.headers['set-cookie'] === undefined) {
						throw new Error(`login did not return with session cookie`);
					    }
					    cookie = { headers: { 'Cookie': resp.response.headers['set-cookie'], withCredentials: true } };
					    return axios.get('http://localhost:8080/', cookie);
					})
				    .then((resp) => {
					    if (resp.data !== 'totopouet') {
						throw new Error(`unexpected payload querying service having logged in`);
					    }
					    return axios.post('http://localhost:8080/logout', cookie, { maxRedirects: 0 })
						.then((e) => done(e || 'unexpected error'))
						.catch((conf) => {
							httpHandler.close();
							if (conf.response === undefined || conf.response.headers === undefined) {
							    done(`headers missing serving logout`);
							} else if (conf.response.headers['set-cookie'] !== undefined) {
							    done(`session still shows after logout`);
							} else { done(); }
						    });
					})
				    .catch((e) => done(e || `undefined error confirming login succeeded`));
			    })
			.catch((e) => done(e || `undefined error fetching app index`));
		});
	}
    });


index.js
'use strict';

const logger = require('wraplog')('sessions-store');

class sessionStore {
    constructor (opts, session) {
	try {
	    opts = opts || { driver: 'sqlite' };
	    let sessionStore = false;
	    let storeOptions = false;
	    if (opts.remoteArray === undefined) {
		if (opts.host !== undefined) { opts.remoteArray = [ opts.host ]; }
		else { opts.remoteArray = [ '127.0.0.1' ]; }
	    }
	    if (opts.driver === undefined || opts.driver === 'sqlite') {
		sessionStore = require('connect-sqlite3')(session);
		storeOptions = {
			db: opts.database || 'test.sqlite',
			dir: opts.path || './',
			table: opts.table || 'sessions'
		    };
	    } else if (opts.driver === 'cassandra') {
		sessionStore = require('cassandra-session-store');
		storeOptions = {
			clientOptions: {
				contactPoints: opts.remoteArray,
				keyspace: opts.database || 'sessions',
				queryOptions: { prepare: true }
			    },
			table: opts.table || 'sessions'
		    };
	    } else if (opts.driver === 'memcached') {
		sessionStore = require('connect-memcached')(session);
		storeOptions = {
			hosts: opts.remoteArray || [ '127.0.0.1' ],
			prefix: opts.prefix || 'sessions:',
			ttl: opts.expiration || 86400
		    };
		if (opts.secret !== undefined) { storeOptions.secret = opts.secret; }
	    } else if (opts.driver === 'mongodb') {
		sessionStore = require('connect-mongodb-session')(session);
		opts.host = opts.host || '127.0.0.1';
		opts.port = opts.port || 27000;
		opts.database = opts.database || 'sessions';
		storeOptions = {
			collection: opts.table || 'sessions',
			expires: opts.expiration || 86400,
			ttl: opts.expiration || 86400,
			uri: `mongodb://${opts.host}:${opts.port}/${opts.database}`
		   };
	    } else if (opts.driver === 'mysql') {
		sessionStore = require('express-mysql-session')(session);
		storeOptions = {
			connectionLimit: opts.connectionLimit || 5,
			createDatabaseTable: opts.createTable || true,
			database: opts.database || 'sessions',
			expiration: opts.expiration || 86400,
			host: opts.host || 'localhost',
			port: opts.port || 3306,
			schema: { tableName: opts.table || 'sessions' }
		};
		if (opts.username !== undefined) { storeOptions.user = opts.username; }
		if (opts.password !== undefined) { storeOptions.password = opts.password; }
	    } else if (opts.driver === 'postgres') {
		sessionStore = require('connect-pg-simple')(session);
		opts.host = opts.host || '127.0.0.1';
		opts.port = opts.port || 5432;
		opts.database = opts.database || 'sessions';
		if (opts.authstr === undefined) {
		    if (opts.password !== undefined && opts.username !== undefined) {
			opts.authstr = `${opts.username}:${opts.password}@`;
		    } else if (opts.username !== undefined) {
			opts.authstr = `${opts.username}:@`;
		    } else { opts.authstr = ''; }
		}
		storeOptions = {
			conString: `postgresql://${opts.authstr}${opts.host}:${opts.port}/${opts.database}`,
			tableName: opts.table || 'sessions', //!!singular
			ttl: opts.expiration || 86400,
		    };
	    } else if (opts.driver === 'redis') {
		sessionStore = require('connect-redis')(session);
		storeOptions = {
			db: opts.database || 0,
			host: opts.host || '127.0.0.1',
			port: opts.port || 6379,
			prefix: opts.prefix || 'sessions:',
			ttl: opts.expiration || 86400,
			logErrors: logger.error
		    };
		if (opts.password !== undefined) { storeOptions.auth_pass = opts.password; }
	    } else if (opts.driver === 'aerospike') {
		sessionStore = require('aerospike-session-store')(session);
		opts.host = opts.remoteArray.join(',');
		storeOptions = {
			namespace: opts.database || 'sessions',
			set: opts.table || 'sessions',
			ttl: opts.expiration || 86400,
			hosts: opts.host || '127.0.0.1:3000'
		    };
	    } else { throw new Error('unhandled opts', opts); }
	    this._sessionStore = new sessionStore(storeOptions);
	    this._sessionStore.on('error', (e) => {
		    logger.error('caught:', JSON.stringify(e || 'undefined error'));
		});
	    logger.info(`${opts.driver} connector instantiated`);
	} catch(e) {
	    logger.error('failed initializing session store with', e !== undefined ? JSON.stringify(e) : 'undefined error');
	    process.exit(1);
	}
    }

    get sessionStoreInstance() {
	return this._sessionStore;
    }
}

module.exports = (opts, session) => new sessionStore(opts, session).sessionStoreInstance;
